File: openresty/lua/dictionary.lua
-- Warm dictionary: colon words persist in dictionary_dir/words/*.fs.
-- Matches harness/src/livingdict/dictionary.py.

local hostmod = require("host")

local M = {}

local RESERVED = {
  ["READ-FILE"] = true,
  ["LIST-DIR"] = true,
  SEARCH = true,
  ["WRITE-FILE"] = true,
  ["RUN-TESTS"] = true,
  ["RUN-GATES"] = true,
  RECEIPT = true,
  ["USE-ARTIFACT"] = true,
  DUP = true,
  DROP = true,
  SWAP = true,
  OVER = true,
  ["+"] = true,
  ["-"] = true,
  ["*"] = true,
  IF = true,
  ELSE = true,
  THEN = true,
  [":"] = true,
  [";"] = true,
}

function M.safe_name(name)
  return type(name) == "string" and name:match("^[A-Z][A-Z0-9-]*$") ~= nil and #name <= 63
end

function M.words_dir(dictionary_dir)
  if not dictionary_dir or dictionary_dir == "" then
    return nil
  end
  return dictionary_dir .. "/words"
end

function M.load_prelude(dictionary_dir)
  local dir = M.words_dir(dictionary_dir)
  if not dir or not hostmod.is_dir(dir) then
    return ""
  end
  local chunks = {}
  for _, item in ipairs(hostmod.walk_files(dir, {})) do
    if item.rel:match("%.fs$") then
      local text = hostmod.read_bytes(item.full)
      if text then
        text = text:gsub("^%s+", ""):gsub("%s+$", "")
        if text ~= "" then
          chunks[#chunks + 1] = { rel = item.rel, text = text }
        end
      end
    end
  end
  table.sort(chunks, function(a, b)
    return a.rel < b.rel
  end)
  local out = {}
  for i = 1, #chunks do
    out[i] = chunks[i].text
  end
  return table.concat(out, "\n")
end

function M.compose(prelude, program)
  prelude = (prelude or ""):gsub("^%s+", ""):gsub("%s+$", "")
  program = (program or ""):gsub("^%s+", ""):gsub("%s+$", "")
  if prelude == "" then
    return program
  end
  if program == "" then
    return prelude
  end
  return prelude .. "\n" .. program
end

function M.loaded_names(dictionary_dir)
  local dir = M.words_dir(dictionary_dir)
  if not dir or not hostmod.is_dir(dir) then
    return {}
  end
  local names = {}
  for _, item in ipairs(hostmod.walk_files(dir, {})) do
    local stem = item.rel:match("([^/]+)%.fs$")
    if stem then
      stem = string.upper(stem)
      if M.safe_name(stem) then
        names[#names + 1] = stem
      end
    end
  end
  table.sort(names)
  return names
end

function M.tokens_to_source(tokens)
  local parts = {}
  for i = 1, #tokens do
    local token = tokens[i]
    if token.kind == "string" then
      parts[#parts + 1] = 'S" ' .. tostring(token.value) .. '"'
    elseif token.kind == "number" then
      parts[#parts + 1] = tostring(token.value)
    else
      parts[#parts + 1] = tostring(token.value)
    end
  end
  return table.concat(parts, " ")
end

function M.used_names(program, loaded)
  local wanted = {}
  for i = 1, #(loaded or {}) do
    wanted[string.upper(tostring(loaded[i]))] = true
  end
  if next(wanted) == nil then
    return {}
  end
  local forth = require("forth")
  local ok, tokens = pcall(forth.tokenize, program)
  if not ok then
    return {}
  end
  local used, seen = {}, {}
  for i = 1, #tokens do
    local token = tokens[i]
    if token.kind == "word" then
      local name = string.upper(tostring(token.value))
      if wanted[name] and not seen[name] then
        seen[name] = true
        used[#used + 1] = name
      end
    end
  end
  table.sort(used)
  return used
end

function M.save_colon(dictionary_dir, vm, objects_root)
  local dir = M.words_dir(dictionary_dir)
  if not dir or type(vm) ~= "table" or type(vm.colon) ~= "table" then
    return {}
  end
  hostmod.mkdir_p(dir)
  local written = {}
  for name, body in pairs(vm.colon) do
    local key = string.upper(tostring(name))
    if M.safe_name(key) and not RESERVED[key] then
      local src
      if body and #body > 0 then
        src = ": " .. key .. " " .. M.tokens_to_source(body) .. " ;\n"
      else
        src = ": " .. key .. " ;\n"
      end
      if objects_root and objects_root ~= "" then
        hostmod.intern(objects_root, src)
      end
      local path = dir .. "/" .. key .. ".fs"
      local existing = hostmod.read_bytes(path)
      if existing ~= src then
        hostmod.write_bytes(path, src)
        written[#written + 1] = key
      end
    end
  end
  table.sort(written)
  return written
end

return M

File: harness/src/livingdict/dictionary.py
"""Warm dictionary: colon words that persist across turns.

The product lives in the workspace. Harness skills live in
`dictionary_dir/words/*.fs` as `: NAME ... ;` sources. They are
prepended to the next episode so the critic and the VM both see them.
"""

from __future__ import annotations

import re
from pathlib import Path
from typing import Any, Iterable

from .forth import Token, tokenize

SAFE_NAME = re.compile(r"^[A-Z][A-Z0-9-]{0,62}$")

RESERVED = frozenset(
    {
        "READ-FILE",
        "LIST-DIR",
        "SEARCH",
        "WRITE-FILE",
        "RUN-TESTS",
        "RUN-GATES",
        "RECEIPT",
        "USE-ARTIFACT",
        "DUP",
        "DROP",
        "SWAP",
        "OVER",
        "+",
        "-",
        "*",
        "IF",
        "ELSE",
        "THEN",
        ":",
        ";",
    }
)


def words_dir(dictionary_dir: str | Path | None) -> Path | None:
    if dictionary_dir is None or str(dictionary_dir) == "":
        return None
    return Path(dictionary_dir) / "words"


def load_prelude(dictionary_dir: str | Path | None) -> str:
    root = words_dir(dictionary_dir)
    if root is None or not root.is_dir():
        return ""
    chunks: list[str] = []
    for path in sorted(root.glob("*.fs")):
        try:
            text = path.read_text(encoding="utf-8").strip()
        except OSError:
            continue
        if text:
            chunks.append(text)
    return "\n".join(chunks)


def compose_program(prelude: str, program: str) -> str:
    prelude = prelude.strip()
    program = program.strip()
    if not prelude:
        return program
    if not program:
        return prelude
    return prelude + "\n" + program


def loaded_names(dictionary_dir: str | Path | None) -> list[str]:
    root = words_dir(dictionary_dir)
    if root is None or not root.is_dir():
        return []
    return sorted(path.stem.upper() for path in root.glob("*.fs") if SAFE_NAME.match(path.stem.upper()))


def tokens_to_source(tokens: Iterable[Token]) -> str:
    parts: list[str] = []
    for token in tokens:
        if token.kind == "string":
            parts.append(f'S" {token.value}"')
        elif token.kind == "number":
            parts.append(str(token.value))
        else:
            parts.append(str(token.value))
    return " ".join(parts)


def used_names(program: str, names: Iterable[str]) -> list[str]:
    wanted = {str(name).upper() for name in names}
    if not wanted:
        return []
    try:
        tokens = tokenize(program)
    except Exception:
        return []
    seen: set[str] = set()
    used: list[str] = []
    for token in tokens:
        if token.kind != "word":
            continue
        name = str(token.value).upper()
        if name in wanted and name not in seen:
            seen.add(name)
            used.append(name)
    return used


def save_colon_words(
    dictionary_dir: str | Path | None,
    colon: dict[str, list[Token]],
    store: Any | None = None,
) -> list[str]:
    root = words_dir(dictionary_dir)
    if root is None:
        return []
    root.mkdir(parents=True, exist_ok=True)
    written: list[str] = []
    for name, body in sorted(colon.items()):
        key = str(name).upper()
        if key in RESERVED or not SAFE_NAME.match(key):
            continue
        source = f": {key} {tokens_to_source(body)} ;\n" if body else f": {key} ;\n"
        target = root / f"{key}.fs"
        data = source.encode("utf-8")
        if store is not None:
            store.intern(data)
        try:
            if target.is_file() and target.read_bytes() == data:
                continue
            target.write_bytes(data)
        except OSError:
            continue
        written.append(key)
    return written
